pgxn/neon/console_connector.c

```diff
@@ -0,0 +1,632 @@
+{
+    JsonbValue k, v;
+    *estimated_len += strlen(key) + strlen(value) + 6; // Add 6 because of two pairs of quotes, a comma, and a colon
+    k.type = jbvString;
+    k.val.string.len = strlen(key);
+    k.val.string.val = key;
+    v.type = jbvString;
+    v.val.string.len = strlen(value);
+    v.val.string.val = value;
+    pushJsonbValue(state, WJB_KEY, &k);
+    pushJsonbValue(state, WJB_VALUE, &v);
+}
+
+static char *ConstructDeltaMessage()
+{
+    JsonbParseState *state = NULL;
+    pushJsonbValue(&state, WJB_BEGIN_ARRAY, NULL);
+    int estimated_len = 0;
+    if(RootTable.db_table)
+    {
+        HASH_SEQ_STATUS status;
+        DbEntry *entry;
+        hash_seq_init(&status, RootTable.db_table);
+        while((entry = hash_seq_search(&status)) != NULL)
+        {
+            pushJsonbValue(&state, WJB_BEGIN_OBJECT, NULL);
+            PushKeyValue(&state, &estimated_len, "op", entry->type == Op_Set ? "set" : "del");
+            PushKeyValue(&state, &estimated_len, "name", entry->name);
+            PushKeyValue(&state, &estimated_len, "type", "db");
+            if(entry->owner[0] != '\0')
+            {
+                PushKeyValue(&state, &estimated_len, "owner", entry->owner);
+            }
+            if(entry->old_name[0] != '\0')
+            {
+                PushKeyValue(&state, &estimated_len, "old_name", entry->old_name);
+            }
+            pushJsonbValue(&state, WJB_END_OBJECT, NULL);
+            estimated_len += 3; // Add 3 for two curly braces and a comma
+        }
+    }
+
+    if(RootTable.role_table)
+    {
+        HASH_SEQ_STATUS status;
+        RoleEntry *entry;
+        hash_seq_init(&status, RootTable.role_table);
+        while((entry = hash_seq_search(&status)) != NULL)
+        {
+            pushJsonbValue(&state, WJB_BEGIN_OBJECT, NULL);
+            PushKeyValue(&state, &estimated_len, "op", entry->type == Op_Set ? "set" : "del");
+            PushKeyValue(&state, &estimated_len, "name", entry->name);
+            PushKeyValue(&state, &estimated_len, "type", "role");
+            if(entry->password)
+            {
+                PushKeyValue(&state, &estimated_len, "password", (char *)entry->password);
+            }
+            if(entry->old_name[0] != '\0')
+            {
+                PushKeyValue(&state, &estimated_len, "old_name", entry->old_name);
+            }
+            pushJsonbValue(&state, WJB_END_OBJECT, NULL);
+            estimated_len += 3;
+        }
+    }
+    JsonbValue *result = pushJsonbValue(&state, WJB_END_ARRAY, NULL);
+    Jsonb *jsonb = JsonbValueToJsonb(result);
+    return JsonbToCString(NULL, &jsonb->root, estimated_len);
+}
+
+static void SendDeltasToConsole()
+{
+    if(!RootTable.db_table && !RootTable.role_table)
+        return;
+    if(!ConsoleURL)
+    {
+        elog(LOG, "ConsoleURL not set, skipping forwarding");
+        return;
+    }
+    CURL *curl = curl_easy_init();
```

Comment:
Some care is needed to not leak this. If any of the elog(ERROR) codepaths will cause a "longjmp" to the error handling routine which will abort the transaction and release all memory contexts (except for TopTransactionContext etc. long lived ones) and all resources tracked by ResourceOwners. Postgres uses ResourceOwners to close things like file descriptors on abort.

You could use the RegisterResourceReleaseCallback() interface to use the ResourceOwner mechanism to clean this up on error. It's a bit tedious to from extensions, though. See `contrib/pgcrypto/openssl.c` for an example, it uses RegisterResourceReleaseCallback() to track openssl handles.

Another approach would be to use PG_TRY()/PG_CATCH(). See `xmlelement` in contrib/xml/xml.c for one example of that. I think that will be easier than RegisterResourceReleaseCallback().

Reply:
Note also that any memory allocation with palloc() etc. will also "elog(ERROR)" on out-of-memory. So it is not feasible/sufficient to just add `curl_easy_cleanup` before every `elog(ERROR)`.

Reply:
Yet another approach would be to store the curl handle in a global variable. Create it the first time it's needed, and keep it around forever after that. Then you don't need to clean it up on error.

Reply:
> You could use the RegisterResourceReleaseCallback() interface to use the ResourceOwner mechanism to clean this up on error. It's a bit tedious to from extensions, though. See `contrib/pgcrypto/openssl.c` for an example, it uses RegisterResourceReleaseCallback() to track openssl handles.

Heh, I remembered I actually did just that in Greenplum some years ago, also for curl handles: https://github.com/greenplum-db/gpdb/blob/main/src/backend/access/external/url_curl.c :smile:. You could copy that, it's Apache 2 licensed like Neon, just make sure to keep the copyright notices intact. 

However, I think the PG_TRY/CATCH or the global variable and reuse the handle approach would be simpler in this case.

Reply:
Switched to the global

Reply:
I was also worried about the ContentHeader linked list that needs to be allocated, so I just moved that to a global as well that gets allocated once on startup